src/proxy/dns.rs

```diff
@@ -0,0 +1,1164 @@
+use istio_dns::proxy::Proxy;
+use istio_dns::resolver::{Answer, Resolver};
+use itertools::Itertools;
+use log::warn;
+use once_cell::sync::Lazy;
+use rand::seq::SliceRandom;
+use rand::thread_rng;
+use std::collections::HashSet;
+use std::net::{IpAddr, SocketAddr};
+use std::ops::Deref;
+use std::str::FromStr;
+use std::sync::{Arc, RwLock};
+use std::time::Duration;
+use tokio::net::{TcpListener, UdpSocket};
+use trust_dns_proto::error::ProtoErrorKind;
+use trust_dns_proto::op::ResponseCode;
+use trust_dns_proto::rr::{Name, RData, Record, RecordType};
+use trust_dns_resolver::system_conf::read_system_conf;
+use trust_dns_server::authority::LookupError;
+use trust_dns_server::server::Request;
+use trust_dns_server::ServerFuture;
+
+const DEFAULT_TCP_REQUEST_TIMEOUT: u64 = 5;
+const DEFAULT_TTL_SECONDS: u32 = 30;
+
+static SVC_CLUSTER_LOCAL: Lazy<Name> = Lazy::new(|| Name::from_str("svc.cluster.local").unwrap());
+static CLUSTER_LOCAL: Lazy<Name> = Lazy::new(|| Name::from_str("cluster.local").unwrap());
+
+/// A proxy that serves known hostnames from ztunnel data structures. Unknown hosts are
+/// forwarded to an upstream resolver.
+pub(super) struct DnsProxy {
+    addr: SocketAddr,
+    server: ServerFuture<Proxy>,
+}
+
+impl DnsProxy {
+    /// Creates a new proxy.
+    ///
+    /// # Arguments
+    ///
+    /// * `addr` - The socket address on which to run the DNS proxy.
+    /// * `network` - The network of the current node.
+    /// * `state` - The state of this proxy.
+    /// * `forwarder` - The forwarder to use for requests not handled by this proxy.
+    pub(super) async fn new<S: AsRef<str>>(
+        addr: SocketAddr,
+        network: S,
+        state: Arc<RwLock<ProxyState>>,
+        forwarder: Arc<dyn Forwarder>,
+    ) -> Result<Self, Error> {
+        // Create the DNS server, backed by ztunnel data structures.
+        let handler = Proxy::new(Arc::new(DnsStore {
+            state,
+            network: network.as_ref().to_string(),
+            forwarder,
+        }));
+        let mut server = ServerFuture::new(handler);
+
+        // Bind and register the UDP socket.
+        let udp_socket = UdpSocket::bind(addr)
+            .await
+            .map_err(|e| Error::Bind(addr, e))?;
+        server.register_socket(udp_socket);
+
+        // Bind and register the TCP socket.
+        let tcp_listener = TcpListener::bind(addr)
+            .await
+            .map_err(|e| Error::Bind(addr, e))?;
+        server.register_listener(
+            tcp_listener,
+            Duration::from_secs(DEFAULT_TCP_REQUEST_TIMEOUT),
+        );
+
+        Ok(Self { addr, server })
+    }
+
+    /// Returns the address to which this DNS proxy is bound.
+    pub(super) fn address(&self) -> SocketAddr {
+        self.addr
+    }
+
+    /// Runs this DNS proxy to completion.
+    pub async fn run(self) {
+        // TODO(nmittler): Do we need to use drain?
+        if let Err(e) = self.server.block_until_done().await {
```

Comment:
What marks it "done"? Do wen eed to close the listeners ourself?

Reply:
Seems we get an error on sigterm:

```
WARN ztunnel::proxy::dns: DNS server shutdown error: Internal error in spawn: task 4 was cancelled
```

Reply:
Yeah, probably worth upstreaming a fix a fix here. I'm guessing the behavior we want would be similar to our draining logic?

Reply:
Maybe yeah. It probably matters less for DNS though since there is less persistent connections (none for UDP)

Reply:
Raised an issue with Trust-DNS: https://github.com/bluejekyll/trust-dns/issues/1976. Let me know if the ask sounds right to you.